File: BOJ/2042. 구간 합 구하기/solution.js
class FenwickTree {
    constructor(size) {
        this.tree = Array(size + 1).fill(0n);
    }

    init(arr) {
        for (let i = 0; i < arr.length; i += 1) {
            this.update(i + 1, arr[i]);
        }
    }

    update(pos, val) {
        while (pos < this.tree.length) {
            this.tree[pos] += val;
            pos += (pos & -pos);
        }
    }

    sum(pos) {
        let ret = 0n;
        while (pos > 0) {
            ret += this.tree[pos];
            pos -= (pos & -pos);
        }
        return ret;
    }
}

const [info, ...data] = require('fs').readFileSync('./data.in').toString().trim().split('\n');
const [n, m, k] = info.split(' ').map(Number);
const nums = data.slice(0, n).map(BigInt);
const queries = data.slice(n).map(d => d.split(' ').map(Number));

const ft = new FenwickTree(n);
ft.init(nums);

const res = queries.reduce((acc, [a, b, c]) => {
    if (a === 1) {
        const u = BigInt(c);
        ft.update(b, u - nums[b - 1]);
        nums[b - 1] = u;
    }
    else {
        acc += (ft.sum(c) - ft.sum(b - 1)).toString() + '\n';
    }
    return acc;
}, '');

console.log(res);

File: BOJ/2042. 구간 합 구하기/solution.cc
#include <iostream>

#define MAX 1000001

using namespace std;

typedef long long ll;

ll tree[MAX], nums[MAX];

int n, m, k, a;
ll b, c;

void update(int pos, ll val) {
    while (pos <= n) {
        tree[pos] += val;
        pos += (pos & -pos);
    }
}

ll sum(int pos) {
    ll ret = 0;
    while (pos > 0) {
        ret += tree[pos];
        pos -= (pos & -pos);
    }
    return ret;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    cin >> n >> m >> k;
    for (int i = 1; i <= n; i += 1) {
        cin >> nums[i];
        update(i, nums[i]);
    }

    for (int i = 0; i < m + k; i += 1) {
        cin >> a >> b >> c;
        if (a == 1) {
            ll d = c - nums[b];
            nums[b] = c;
            update(b, d);
        }
        else {
            cout << sum(c) - sum(b - 1) << '\n';
        }
    }

    return 0;
}